clients/csrf/default_csrf_token_manager.go

```diff
@@ -2,100 +2,40 @@ package csrf
 
 import (
 	"net/http"
+	"time"
 )
 
-const XCsrfHeader = "X-Csrf-Header"
-const XCsrfToken = "X-Csrf-Token"
-const CsrfTokenHeaderRequiredValue = "Required"
+const CsrfSessionTimeout = 2 * time.Minute
 
 type DefaultCsrfTokenManager struct {
-	request          *http.Request
-	transport        *Transport
+	csrf             *Csrf
 	csrfTokenFetcher CsrfTokenFetcher
 }
 
-func NewDefaultCsrfTokenManager(transport *Transport, request *http.Request) *DefaultCsrfTokenManager {
-	return &DefaultCsrfTokenManager{request: request, transport: transport, csrfTokenFetcher: NewDefaultCsrfTokenFetcher(transport)}
+func NewDefaultCsrfTokenManager(transport *Transport) *DefaultCsrfTokenManager {
+	return &DefaultCsrfTokenManager{csrf: transport.Csrf, csrfTokenFetcher: NewDefaultCsrfTokenFetcher(transport)}
 }
 
-func NewDefaultCsrfTokenManagerWithFetcher(transport *Transport, request *http.Request, csrfTokenFetcher CsrfTokenFetcher) *DefaultCsrfTokenManager {
-	return &DefaultCsrfTokenManager{request: request, transport: transport, csrfTokenFetcher: csrfTokenFetcher}
+func NewDefaultCsrfTokenManagerWithFetcher(csrf *Csrf, csrfTokenFetcher CsrfTokenFetcher) *DefaultCsrfTokenManager {
+	return &DefaultCsrfTokenManager{csrf: csrf, csrfTokenFetcher: csrfTokenFetcher}
 }
 
-func (c *DefaultCsrfTokenManager) updateToken() error {
-	if !c.shouldInitialize() {
+func (c *DefaultCsrfTokenManager) updateToken(req *http.Request) error {
+	if !c.csrf.IsProtectionRequired(req) {
 		return nil
 	}
-	err := c.initializeToken(false)
-	if err != nil {
-		return err
-	}
-
-	c.updateTokenInRequest()
 
-	return nil
-}
+	c.csrf.Mutex.Lock()
+	defer c.csrf.Mutex.Unlock()
 
-func (c *DefaultCsrfTokenManager) initializeToken(forceInitializing bool) error {
-	if forceInitializing || !c.transport.Csrf.IsInitialized {
-		var err error
-		csrfToken, err := c.csrfTokenFetcher.FetchCsrfToken(getCsrfTokenUrl(c.request), c.request)
-		if csrfToken == nil {
-			return nil
-		}
-		c.transport.Csrf.Header, c.transport.Csrf.Token = csrfToken.CsrfTokenHeader, csrfToken.CsrfTokenValue
+	if c.csrf.IsExpired(CsrfSessionTimeout) {
```

Comment:
I think this is brittle. 
Yes, the session timeout is 2 minutes and the token is per session, but there is some time lost during routing, server processing of the returned token, etc. There might be a case when the token has less than 2 minutes of expiration left.
I'd say 1:30 minutes max. We've seen some fetch token requests that took 10 seconds and nobody knows where exactly was the delay and when the token/session was created

Reply:
The less the safer

Reply:
Reduced to 1m30s